dynamic_instrumentation/src/source.rs

```diff
@@ -0,0 +1,67 @@
+use c2rust_analysis_rt::mir_loc::MirPlace;
+use rustc_middle::mir::{Local, Operand, Place, Rvalue};
+
+use crate::util::Convert;
+
+pub trait Source {
+    fn source(&self) -> Option<MirPlace>;
+}
+
+impl Source for Place<'_> {
+    fn source(&self) -> Option<MirPlace> {
+        Some(self.convert())
+    }
+}
+
+impl Source for Operand<'_> {
+    fn source(&self) -> Option<MirPlace> {
+        self.place().as_ref().and_then(Place::source)
+    }
+}
+
+impl Source for Vec<Operand<'_>> {
+    fn source(&self) -> Option<MirPlace> {
+        // TODO: have hook-specific sources
+        self.first().and_then(Operand::source)
+    }
+}
+
+impl Source for Rvalue<'_> {
+    fn source(&self) -> Option<MirPlace> {
+        rv_place(self).as_ref().and_then(Place::source)
+    }
+}
+
+impl Source for Local {
+    fn source(&self) -> Option<MirPlace> {
+        Place::from(*self).source()
+    }
+}
+
+impl Source for u32 {
+    fn source(&self) -> Option<MirPlace> {
+        Local::from_u32(*self).source()
+    }
+}
+
+/// Get the one and only input [`Place`], if applicable.
+fn rv_place<'tcx>(rv: &'tcx Rvalue) -> Option<Place<'tcx>> {
```

Comment:
`mir_utils`

Reply:
`rv_value` is only used here by a `Source` `impl`, so it's private and encapsulated.  I think it should stay together.

Reply:
you're right it's only used here, but that's coincidental -- we have a bunch of MIR-specific utilities and this can be one of them, there is no issue making it public (it's similar to `Operand::place` anyhow)

Reply:
Okay.